### DISCUSSION TITLE
Help me understand Avoid Mapping Selectors

### DISCUSSION THREAD

**Discussion by stocksr** [timdeschryver/eslint-plugin-ngrx#288] [Q&A]
so I have some logic like this in an effect
```ts
        return this.store.select(selectSomeBitOfState).pipe(
            filter((s) => s.status === 'busy'),
            map((s) => s.query),
            distinctUntilChanged(),
            switchMap((q)=>{ /* some logic to run the query */)
```
         
the rule    `ngrx/avoid-mapping-selectors` is objecting to the map 

what is the recommended way of implementing this?

Specifically  - I need to run my logic for the latest value of query when the status is 'busy' all other status should be ignored

p.s. SwitchMap is correct as if the query changes while I am processing I do want to restart the operation.

**Comment by timdeschryver** [comment]
Hi @stocksr , good question!
It might be that we've made it too strict, see #285 for another example.
Are you doing this in an effect or in the component?
An alternative way is to do the distinct differently, for example with `distinctUntilKeyChanged('query')`

**Reply by stocksr** [reply]
> Hi @stocksr , good question! It might be that we've made it too strict, see #285 for another example. Are you doing this in an effect or in the component? An alternative way is to do the distinct differently, for example with `distinctUntilKeyChanged('query')`


Hi @timdeschryver, thanks for the reply. I think I understand your suggestion: you are suggesting that rather than have the map call after the filter - I simply pass the unmapped object down the pipeline and re-write the rest of the pipeline to access the property of the unmapped object. So code wise remove `map(x=>x.prop)` and in the rest of the pipeline use `x.prop`. 

I am trying to understand the why here, could you please explain the justification for making this change to my code, is there a run time performance improvement to this, or is there a developer productivity / safety gain.

**Reply by timdeschryver** [reply]
@stocksr Yea, correct. I only mentioned it as a workaround for this rule.
As I said, it could be that the current implementation of the rule is too strict.

To have more context, is this code snippet used in an effect, component, guard, ...?

**Reply by stocksr** [reply]
> To have more context, is this code snippet used in an effect, component, guard, ...?

Sorry - in this case this is in an effect: the entire effect looks a bit like this - edits to remove names specific to my project

```ts
    public runQuery$ = createEffect(() => {
        return this.store.select(selectSomeBitOfState).pipe(
            filter((s) => s.status === 'busy'),
            map((s) => s.query),

            distinctUntilChanged(),

            switchMap((q) =>
                this.myService
                    .makeAHttpPostMethodToTheQueryEndpoint({
                        body: {
                            searchTerms: q.search,
                            pageNumber: q.page,
                            filter: q.filter
                        },
                    })
                    .pipe(
                        map((d) => querySuccessAction({ response: d })),
                        catchError((err: unknown, c) => {
                            console.warn('err', err, c);
                            return of(queryErrorAction({ error: err }));
                        })
                    )
            )
        );
    });

```

**Reply by timdeschryver** [reply]
Cool, thanks 👍
I think that allowing this pattern in an effect is a valid use case, and I've bumped against a similar issue not too long ago. I'll update the rule to allow this.

**Reply by timdeschryver** [reply]
Should be resolved with https://github.com/timdeschryver/eslint-plugin-ngrx/pull/289
